<a href="https://colab.research.google.com/github/Justine-Villacin/CPE-103-OOP-1-A/blob/main/Laboratory_No_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
# distance is a class. Distance is measured in terms of feet and inches class distance:
class distance:
    def __init__(self, f, i):
      self.feet = f
      self.inches = i

    # overloading of binary operator > to compare two distances
    def __gt__(self, other):
      if (self.feet > other.feet):
        return True
      elif (self.feet == other.feet) and (self.inches > other.inches):
        return True
      else:
        return False

    # overloading of binary operator + to add two distances
    def __add__(self, other):
      f = self.feet + other.feet
      i = self.inches + other.inches
      if (i >= 12):
        i -= 12
        f += 1
      return distance(f, i)

    # displaying the distance
    def show(self):
      print("Feet = ", self.feet, "Inches = ", self.inches)

a, b = (input("Enter feet and inches of distance 1: ")).split()
a = int(a)
b = int(b)
c, d = (input("Enter feet and inches of distance 2: ")).split()
c = int(c)
d = int(d)
d1 = distance(a, b)
d2 = distance(c, d)

if (d1 > d2):
    print("Distance 1 is greater than distance 2")
else:
    print("Distance 2 is greater than distance 1")

d3 = d1 + d2
print("The sum of the distances is: ")
d3.show()

Enter feet and inches of distance 1: 3 4
Enter feet and inches of distance 2: 5 6
Distance 2 is greater than distance 1
The sum of the distances is: 
Feet =  8 Inches =  10


In [23]:
class RegularPolygon:
    def __init__ (self, side):
      self ._side = side
class Square (RegularPolygon):
    def area (self):
      return self._side * self._side
class EquilateralTriangle (RegularPolygon):
    def area (self):
      return self._side * self._side * 0.433

obj1 = Square(4)
obj2 = EquilateralTriangle(3)

print (obj1.area())
print (obj2.area())


16
3.897


# Observation

> My Observation in this program is the given class in this was regularpolygon or parentclass then need to calculate the area of the regularpolygon given by assigning a sub class of area in everypolygon.



# Supplementary

In [26]:
import math

# Define RegularPolygon class
class RegularPolygon:
    def __init__(self, side):
        self._side = side

# Heptagon class (7 sides)
class Heptagon(RegularPolygon):
    def area(self):
        return (7 / 4) * (self._side ** 2) * (1 / math.tan(math.pi / 7))

# Nonagon class (9 sides)
class Nonagon(RegularPolygon):
    def area(self):
        return (9 / 4) * (self._side ** 2) * (1 / math.tan(math.pi / 9))

# Decagon class (10 sides)
class Decagon(RegularPolygon):
    def area(self):
        return (5 / 2) * (self._side ** 2) * (1 / math.tan(math.pi / 10))


obj6 = Heptagon(7)
obj7 = Nonagon(9)
obj8 = Decagon(10)


print(f"Area of Heptagon: {obj6.area():.2f}")
print(f"Area of Nonagon: {obj7.area():.2f}")
print(f"Area of Decagon: {obj8.area():.2f}")



Area of Heptagon: 178.06
Area of Nonagon: 500.73
Area of Decagon: 769.42


# Supplementary Questions

> For these 3 regularpolygon I used a 2 decimal point indication.

> Heptagon = Heptagon consist of 7 sides and I used the area of heptagon formula for me to calculate the area of this regularpolygon, after calculating i got 178.06 for the area.

> Nonagon = Nonagon have 9 sides and I used also the formula for this regularpolygon to get the area, after calculating i got 500.73 for the area.

> Decagon = have 10 sides and for us to get the exact area of this regularpolygon I need to used the area formula for this shape, after calculating i got 769.42 for the area.







# Questions 1 - 5

> Polymorphism is important because it allows objects to take on many forms, enabling code to treat different classes uniformly.


> Polymorphism increases code reusability but can be difficult to implement and reduce readability.


> The advantage is handling different file types through a common interface, while the disadvantage is increased complexity in handling file-specific nuances.


> Consider the common interface objects will share and the performance implications of runtime method resolution when implementing polymorphism.


> Polymorphism is used in actual programs to allow a single function to handle different types of objects, simplifying code and making it easier to extend programs.





# Conclusion

> In conclusion, polymorphism enhances code flexibility, but requires careful implementation and consideration of its potential drawbacks.
